### PR TITLE
Customize colors 

### LINKED ISSUE #8: Customize admin section

### LINKED ISSUE #9: Customize colors

### PR DESCRIPTION
**Pull request by taitus**
## References

Closes: #8 #9 

## Objectives

Customize colors:
- Front section
- Admin section

Pending

## Visual Changes

Pending

## Notes

Pending

### CODE REVIEW

File: app/assets/stylesheets/custom.scss [hunk 1]
```diff
@@ -414,3 +449,255 @@ header {
...
+
+// 11. Tables (layout.scss)
+// ----------
+
+table {
+
+  tr {
+
+    &:nth-child(odd) {
+      background: #f9f9f9;
```
Comment (javierm) [inline app/assets/stylesheets/custom.scss]: Typo: should be `$gray` or `$background`.

File: app/assets/stylesheets/custom.scss [hunk 2]
```diff
@@ -414,3 +449,255 @@ header {
     }
   }
 }
+
+input:focus {
+  outline: 0;
+}
+
+body {
+  background: $gray;
```
Comment (javierm) [inline app/assets/stylesheets/custom.scss]: Maybe we can use a `$background` variable and assign it to `$gray`; this way we can change the background by changing one variable (since changing `$gray` would be strange if we change it so the color is not gray :sweat_smile:). IMHO the commit "Override body background for all application pages" would be a good place to do so.

File: app/assets/stylesheets/custom.scss [hunk 3]
```diff
@@ -414,3 +449,255 @@ header {
...
+table {
+
+  tr {
+
+    &:nth-child(odd) {
+      background: #f9f9f9;
+    }
+
+    &:nth-child(even) {
+      background: #fff4f4;
```
Comment (javierm) [inline app/assets/stylesheets/custom.scss]: What's the reason behind choosing this color?

File: app/assets/stylesheets/custom.scss [hunk 4]
```diff
@@ -414,3 +449,255 @@ header {
...
+
+  ul {
+    [class^="icon-"] {
+      color: $black;
+    }
+  }
+
+  li {
+    ul {
+      border-left: 1px solid $sidebar-hover;
```
Comment (javierm) [inline app/assets/stylesheets/custom.scss]: This border is almost invisible :thinking:.

File: app/views/custom/layouts/_admin_header.html.erb [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  <div class="expanded row admin-top-bar">
+    <div class="title-bar" data-responsive-toggle="responsive_menu" data-hide-for="medium">
+      <button class="menu-icon" type="button" data-toggle="responsive_menu"></button>
+      <div class="title-bar-title"><%= t("application.menu") %></div>
+    </div>
+
+    <div class="top-bar">
+      <div class="top-bar-title">
+        <% if namespace == "officing" %>
+          <%= link_to "#", accesskey: "0" do %>
```
Comment (javierm) [inline app/views/custom/layouts/_admin_header.html.erb]: Why are we adding an accesskey? :thinking:

I'm not sure about the way accesskeys have evolved in the last few years; in the past they were a bit dangerous because they might interfere with keyboard shortcuts :thinking:.

File: app/views/custom/layouts/_admin_header.html.erb [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+            <%= image_tag(image_path_for("logo_header.png"), class: "float-left", alt: t("layouts.header.logo")) %>
+          <% end %>
+        <% else %>
+          <%= link_to namespaced_root_path, accesskey: "0" do %>
+            <%= image_tag(image_path_for("logo_header.png"), class: "float-left", alt: t("layouts.header.logo")) %>
```
Comment (javierm) [inline app/views/custom/layouts/_admin_header.html.erb]: Use CSS instead of `class: "float-left"` :pray:.

File: app/views/custom/layouts/_admin_header.html.erb [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    </div>
+
+    <div class="top-bar">
+      <div class="top-bar-title">
+        <% if namespace == "officing" %>
+          <%= link_to "#", accesskey: "0" do %>
+            <%= image_tag(image_path_for("logo_header.png"), class: "float-left", alt: t("layouts.header.logo")) %>
+          <% end %>
+        <% else %>
+          <%= link_to namespaced_root_path, accesskey: "0" do %>
```
Comment (javierm) [inline app/views/custom/layouts/_admin_header.html.erb]: Is there a reason why we're removing the `<h1>` tag present in the original file?

File: spec/system/admin/homepage/homepage_spec.rb [hunk 8]
```diff
@@ -22,7 +22,7 @@
     scenario "Admin menu links to homepage path" do
       visit new_admin_widget_card_path(header_card: true)
 
-      click_link Setting["org_name"] + " Administration"
+      click_link("CONSUL logo")
```
Comment (javierm) [inline spec/system/admin/homepage/homepage_spec.rb]: Strange that this test passes :thinking:. I guess we'll have to change the `layouts.header.logo` i18n key so it says the same as what's present in the image :thinking:.

**Review by javierm — COMMENTED** [review]
Initial comments; will continue later.

File: app/assets/stylesheets/custom.scss (hunk 3 above)
Comment (taitus) [reply app/assets/stylesheets/custom.scss]: Because it is the one defined in the mockups 😅 
Captura de pantalla 2020-12-10 a las 16 30 36

File: app/assets/stylesheets/custom.scss (hunk 4 above)
Comment (taitus) [reply app/assets/stylesheets/custom.scss]: Ok, I'll delete this rule

File: app/views/custom/layouts/_admin_header.html.erb (hunk 5 above)
Comment (taitus) [reply app/views/custom/layouts/_admin_header.html.erb:21]: This accesskey is inherited from the layouts/_admin_header.html.erb.
It can also be found in the layouts/_header and in our new custom/layouts/_header.html.erb 😅 

I simply tried to maintain the existing logic in the links and focus on displaying the logo correctly.
But I can remove it if we consider it necessary.

File: app/views/custom/layouts/_admin_header.html.erb (hunk 7 above)
Comment (taitus) [reply app/views/custom/layouts/_admin_header.html.erb]: Because before we showed a text and now an image, that was my motive.
Do you think it's better to keep the h1?

File: app/assets/stylesheets/custom.scss (hunk 3 above)
Comment (javierm) [reply app/assets/stylesheets/custom.scss]: :+1:

Since we've got about 12 usages of the `#fff4f4` color in the `custom.scss` file, how about extracting it to a variable?

File: app/views/custom/layouts/_admin_header.html.erb (hunk 7 above)
Comment (javierm) [reply app/views/custom/layouts/_admin_header.html.erb]: The `<h1>` tag is OK IMHO. Actually CONSUL now uses `<h1>` in the application layout too: consul@bb2e2fa0a.

File: app/assets/stylesheets/custom.scss [hunk 9]
```diff
@@ -247,6 +253,16 @@ header {
...
+
+    .top-bar-title {
+      flex: 19%;
+      margin-right: 0;
+      min-width: 14rem;
+    }
+
+    .responsive-menu {
+      flex: 81%;
+    }
```
Comment (javierm) [inline app/assets/stylesheets/custom.scss]: I see the page the same way with just:

```scss
.top-bar-title {
  margin-right: 0;
}
```

Is there something I'm missing? :thinking:

File: app/assets/stylesheets/custom.scss [hunk 10]
```diff
@@ -414,3 +449,255 @@ header {
...
+.callout {
+  &.primary {
+    background-color: $info-bg;
+    border-color: $info-border;
+    color: $color-info;
+  }
+}
+
+.tabs-content {
+  background: $gray;
```
Comment (javierm) [inline app/assets/stylesheets/custom.scss]: How about overriding `$tab-content-background` instead?

File: app/assets/stylesheets/custom.scss [hunk 11]
```diff
@@ -414,3 +449,255 @@ header {
...
+}
+
+.tabs {
+  background: $gray;
```
Comment (javierm) [inline app/assets/stylesheets/custom.scss]: How about overriding `$tab-background` instead?

File: app/assets/stylesheets/custom.scss [hunk 12]
```diff
@@ -414,3 +449,255 @@ header {
...
+}
+
+.tabs-title.is-active > {
+  [aria-selected='true'] {
+    background: $gray;
```
Comment (javierm) [inline app/assets/stylesheets/custom.scss]: How about overriding `$tab-background-active` instead?

File: app/assets/stylesheets/custom.scss [hunk 13]
```diff
@@ -414,3 +449,255 @@ header {
...
+
+// 04. Stats (admin.scss)
+// ---------
+
+.stats {
+  background: $gray;
+}
+
+.highlight {
+  background: #fff4f4;
```
Comment (javierm) [inline app/assets/stylesheets/custom.scss]: How about overriding `$highlight` instead?

File: app/views/custom/welcome/_header.html.erb [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+        </div>
+      </div>
+
+      <% if header.image.present? %>
+        <div class="small-12 medium-6 column">
+          <%= image_tag(header.image_url(:large),
+                        class: "margin",
+                        alt: header.image.title) %>
+        </div>
+      <% end %>
```
Comment (javierm) [inline app/views/custom/welcome/_header.html.erb]: Instead of copying the view to remove the `background-header` div, do we get the same result adding a `display: none` CSS rule to this element?

**Review by javierm — COMMENTED** [review]
Second batch :smile:.

File: app/assets/stylesheets/custom.scss [hunk 15]
```diff
@@ -414,3 +434,275 @@ header {
...
+  background-color: $background;
+
+  &:hover,
+  &:active,
+  &:focus {
+    background-color: $background;
+  }
+}
+
+input:-webkit-autofill {
```
Comment (javierm) [inline app/assets/stylesheets/custom.scss]: IMHO it isn't necessary to override this style :thinking:, since it isn't defined in CONSUL. Besides, changing the color while the user is selecting an element is probably a usability improvement.

File: app/assets/stylesheets/custom.scss [hunk 16]
```diff
@@ -414,3 +434,275 @@ header {
...
+}
+
+.callout {
+  &.primary {
+    background-color: $info-bg;
+    border-color: $info-border;
+    color: $color-info;
+  }
+}
+
```
Comment (javierm) [inline app/assets/stylesheets/custom.scss]: Are there lines necessary? It seems to me it'll work the same way without them, since we've redefined these variables in custom settings.

File: app/views/layouts/management.html.erb [hunk 17]
```diff
@@ -17,18 +17,11 @@
...
-                  <% end %>
-                </h1>
-              </li>
-            </ul>
-          </div>
+          <div class="top-bar-title">
+            <%= link_to management_root_path, accesskey: "0" do %>
+              <%= image_tag(image_path_for("logo_header.png"), class: "float-left", alt: t("layouts.header.logo")) %>
+            <% end %>
+         </div>
```
Comment (javierm) [inline app/views/layouts/management.html.erb]: Instead of changing the original view, we usually we copy this file to custom, commit, and then add the changes.

By the way, IMHO the changes should be similar to what we've done in `admin_header` (no accesskey, keep the h1 tag, ...)

File: app/assets/stylesheets/custom.scss [hunk 18]
```diff
@@ -414,3 +434,275 @@ header {
...
+    color: $background;
+  }
+}
+
+// 01. Votes and supports (participation.scss)
+// ----------------------
+.proposal .supports {
+
+  .progress {
+    background: lighten($brand, 35%);
```
Comment (javierm) [inline app/assets/stylesheets/custom.scss]: How about overriding the `$proposals` variable instead? :thinking: I think we can remove many changes on this file if we do so.

File: app/assets/stylesheets/custom.scss [hunk 19]
```diff
@@ -414,3 +434,275 @@ header {
...
+
+.key-dates li:hover:not(.is-active) {
+  background: $light-pink;
+}
+
+// 20. Documents
+// -------------
+
+.additional-document-link {
+  background: $light-pink;
```
Comment (javierm) [inline app/assets/stylesheets/custom.scss]: Would we get the same result overriding `$highlight-soft` instead?

File: app/assets/stylesheets/custom.scss [hunk 20]
```diff
@@ -414,3 +434,275 @@ header {
...
+
+// 03. Legislation cards (legislation.scss)
+// -----------------
+
+.legislation-calendar {
+  background: $light-pink;
+}
+
+.key-dates li:hover:not(.is-active) {
+  background: $light-pink;
```
Comment (javierm) [inline app/assets/stylesheets/custom.scss]: No longer necessary since we've overwritten `$highlight`.

File: app/assets/stylesheets/custom.scss [hunk 21]
```diff
@@ -414,3 +434,275 @@ header {
...
+      background: lighten($brand, 25%);
+    }
+  }
+}
+
+// 03. Legislation cards (legislation.scss)
+// -----------------
+
+.legislation-calendar {
+  background: $light-pink;
```
Comment (javierm) [inline app/assets/stylesheets/custom.scss]: No longer necessary since we've overwritten `$highlight`.

File: app/assets/stylesheets/custom.scss [hunk 22]
```diff
@@ -414,3 +434,275 @@ header {
...
+}
+
+// 06. Budgets (participation.scss)
+// -----------
+.groups-and-headings {
+  .heading {
+    a {
+      &:hover {
+        background: $light-pink;
```
Comment (javierm) [inline app/assets/stylesheets/custom.scss]: No longer necessary since we've overwritten `$highlight`.

File: app/assets/stylesheets/custom.scss [hunk 23]
```diff
@@ -414,3 +434,275 @@ header {
...
+
+      &::before {
+        background-color: $brand;
+      }
+    }
+  }
+}
+
+.budgets-stats .header {
+  background: $light-pink;
```
Comment (javierm) [inline app/assets/stylesheets/custom.scss]: No longer necessary since we've overwritten `$highlight`.

File: app/assets/stylesheets/custom.scss [hunk 24]
```diff
@@ -414,3 +434,275 @@ header {
...
+.card {
+  background: $background;
+}
+
+.figure-card figcaption span {
+  background: $background;
+}
+
+select {
+  background-color: $background;
```
Comment (javierm) [inline app/assets/stylesheets/custom.scss]: How about overwriting `$select-background` instead?

File: app/assets/stylesheets/custom.scss [hunk 25]
```diff
@@ -213,6 +210,7 @@ header {
     a {
       font-weight: normal;
       font-size: 1rem;
+      text-decoration: none;
```
Comment (javierm) [inline app/assets/stylesheets/custom.scss]: What's the reason for this change? :thinking:

File: app/assets/javascripts/ckeditor/loader.js.erb [hunk 26]
```diff
@@ -1,3 +1,4 @@
 //= require ckeditor/init
 
 CKEDITOR.config.customConfig = "<%= javascript_path "ckeditor/config.js" %>";
+CKEDITOR.addCss(".cke_editable { background-color: #f9f9f9 }");
```
Comment (javierm) [inline app/assets/javascripts/ckeditor/loader.js.erb]: Isn't it possible to use CSS for this? :thinking:

File: app/assets/stylesheets/custom.scss [hunk 27]
```diff
@@ -414,3 +434,275 @@ header {
...
+
+  .icon-budget {
+    color: $brand;
+  }
+}
+
+.debate-form {
+
+  .recommendations li::before {
+    color: $brand;
```
Comment (javierm) [inline app/assets/stylesheets/custom.scss]: How about overriding `$debates` instead?

File: app/assets/stylesheets/custom.scss [hunk 28]
```diff
@@ -414,3 +434,275 @@ header {
...
+
+.debate-form,
+.proposal-form,
+.budget-investment-form,
+.document-form,
+.topic-new,
+.topic-form {
+
+  .icon-debates {
+    color: $brand;
```
Comment (javierm) [inline app/assets/stylesheets/custom.scss]: How about overriding `$debates` instead?

**Review by javierm — COMMENTED** [review]
Batch three :wink:.

File: app/assets/stylesheets/custom.scss (hunk 25 above)
Comment (taitus) [reply app/assets/stylesheets/custom.scss]: I will remove this change

File: app/assets/javascripts/ckeditor/loader.js.erb (hunk 26 above)
Comment (taitus) [reply app/assets/javascripts/ckeditor/loader.js.erb]: All the documentation I looked at was solved this way. Although I seem to remember that there was some way to load a css and apply those changes(but I think it was always load through the js). I can investigate it if necessary.

File: app/assets/stylesheets/custom.scss [hunk 29]
```diff
@@ -414,3 +433,140 @@ header {
...
+  background: $background;
+}
+
+.notifications .icon-circle {
+  color: $brand;
+}
+
+form {
+  [type]:not([type=submit]):not([type=file]):not([type=checkbox]):not([type=radio]):not(.close-button) {
+    background: $background;
```
Comment (javierm) [inline app/assets/stylesheets/custom.scss]: IMHO it's OK to leave it the CONSUL way, with inputs having a slightly different background color.

File: app/views/custom/layouts/devise.html.erb [hunk 30]
```diff
@@ -0,0 +1,39 @@
+<!DOCTYPE html>
+<html lang="<%= I18n.locale %>">
+  <head>
+    <%= render "layouts/common_head", default_title: setting["org_name"] %>
+    <%= render "layouts/meta_tags" %>
+    <%= raw setting["html.per_page_code_head"] %>
```
[the remark this answers is not in the document]
Comment (javierm) [reply app/views/custom/layouts/devise.html.erb:6]: It's written that way in CONSUL; nothing we can do here.

File: app/views/custom/layouts/devise.html.erb [hunk 31]
```diff
@@ -0,0 +1,39 @@
...
+  </head>
+
+  <body class="auth-page">
+    <%= raw setting["html.per_page_code_body"] %>
```
[the remark this answers is not in the document]
Comment (javierm) [reply app/views/custom/layouts/devise.html.erb:10]: It's written that way in CONSUL; nothing we can do here.

**Review by javierm — APPROVED** [review]
